In this notebook:
    - correlated gauges without predictions (y) to gauges that do have predictions (features)
    - use KNN based on lat, long, alt, basin to determine 5-10 closest gages to those that we are looking for
    - result should be a dataframe with first column as the y USGS gage and each additional column a feature USGS gage that has a NOAA prediction OR a dictionary with each key pointing to the ten closest gages
    - set threshold for how close a gage would need to be to be considered

In [44]:
import pandas as pd
from bs4 import BeautifulSoup
import os
import requests
import time
import re

In [45]:
import pickle
path="C:\Springboard\Github\gauge_info"
os.chdir(path)

In [46]:
# loade the dataframe of potential features that have NOAA predictions
df = pickle.load(open("USGS_features.pkl", "rb"))
df

NOAA_gauge         River State  Elevation Segment  \
0        SPRA3    San Pedro     AZ     2820.0       7   
5        AFHA3    Agua Fria     AZ     4400.0      17   
6        AFMA3    Agua Fria     AZ     3434.0      18   
7        AFRA3    Agua Fria     AZ     1800.0      19   
10       ATPA3   Altar Wash     AZ     2975.0      24   
..         ...           ...   ...        ...     ...   
450      YASC2        Yampa     CO     7240.0       2   
451      YDLC2        Yampa     CO     5600.0      17   
452      YMSC2        Yampa     CO     7050.0       4   
453      YLLU1  Yellowstone     UT     7430.0      20   
454      ZUIN5         Zuni     NM     6450.0       1   

                                         USGS_link      usgs        lat  \
0    http://waterdata.usgs.gov/az/nwis/uv?09472050  09472050  32.446183   
5    http://waterdata.usgs.gov/az/nwis/uv?09512450  09512450  34.485278   
6    http://waterdata.usgs.gov/az/nwis/uv?09512500  09512500  34.315307   
7    http://waterdata.usgs.gov/az/nwis/uv?09512800  09512800  34.015589   
10   http://waterdata.usgs.gov/az/nwis/uv?09486800  09486800  31.838972   
..                                             ...       ...        ...   
450  http://waterdata.usgs.gov/co/nwis/uv?09237450  09237450  40.264261   
451  http://waterdata.usgs.gov/co/nwis/uv?09260050  09260050  40.451634   
452  http://waterdata.usgs.gov/co/nwis/uv?09237500  09237500  40.286544   
453  http://waterdata.usgs.gov/ut/nwis/uv?09292500  09292500  40.511893   
454  http://waterdata.usgs.gov/nm/nwis/uv?09386950  09386950  35.100278   

           long      alt            basin  
0    110.488418  2820.00  Lower San Pedro  
5    112.237500  4400.00        Agua Fria  
6    112.064046  3434.00        Agua Fria  
7    112.167938  1800.00        Agua Fria  
10   111.404269  2975.15     Brawley Wash  
..          ...      ...              ...  
450  106.891767  7240.00      Upper Yampa  
451  108.525101  5600.00      Lower Yampa  
452  106.829056  7050.00      Upper Yampa  
453  110.341549  7430.00         Duchesne  
454  108.751667  6450.00             Zuni  

[373 rows x 11 columns]

In [47]:
# load the dataframe of target gages that we want to model
dt = pickle.load(open("USGS_targets.pkl", "rb"))
dt

USGS        lat        long      alt                 basin
0    10140700  41.231819  111.984497  4285.00           Lower Weber
1    10149000  40.118012  111.314622  6320.00          Spanish Fork
2    10155200  40.554398  111.433243  5691.59                 Provo
3    10224000  39.481898  112.393834  4660.00          Lower Sevier
4    10308200  38.714627  119.764899  5400.00          Upper Carson
..        ...        ...         ...      ...                   ...
147  06892350  38.983337   94.964689   753.87  Lower Kansas, Kansas
148  07169800  37.375597   96.185548   897.30                   Elk
149  06775900  41.778611  100.525278  2798.18                Dismal
150  06461500  42.902222  100.362222  2287.57       Middle Niobrara
151  09076300  39.223200  106.857308  7575.00          Roaring Fork

[125 rows x 5 columns]

One thing we didn't look at in the previous notebook: number of basins that are possible. If we are going to create dummy variables for thie, then we need to know how many dummies. Let's find this value for each DF.

In [48]:
dt['basin'].value_counts()

Upper Rio Grande           7
Roaring Fork               4
Animas                     4
Upper North Platte         4
Clear                      3
                          ..
Lower Sevier               1
Eagle                      1
Clarks Fork Yellowstone    1
Upper Owyhee               1
South Fork Clearwater      1
Name: basin, Length: 84, dtype: int64

In [49]:
df['basin'].value_counts()

Colorado Headwaters                  13
Upper Virgin                         11
Duchesne                             10
Upper Yampa                           9
Upper Weber                           9
                                     ..
Lower Lake Powell                     1
Burro                                 1
Havasu-Mohave Lakes                   1
Lower Colorado-Marble Canyon          1
Lower Gila-Painted Rock Reservoir     1
Name: basin, Length: 121, dtype: int64

So we have 84 basins in the targets list and 121 basins in the features dataframe. We'll probbably have to make dummy variables for these. We may want to consider throwing out gages from the targets if there is not a NOAA gage within that same drainage basin. Let's plot those locations out.

In [50]:
# to make life easier, let's create a column that is consistent with google format (West is negagtive)
df['lng']= -1*df['long']
dt['lng']= -1*dt['long']

In [51]:
from bokeh.io import show
from bokeh.plotting import gmap
from bokeh.models import GMapOptions
from bokeh.models import ColumnDataSource

In [52]:
bokeh_width= 1000
bokeh_height= 800

In [53]:
apikey='AIzaSyChKhMLyzg_dJYipHfbzo1QK7qbPW869WY'
def plot(lat, lng, zoom=10, map_type='roadmap'):
    gmap_options = GMapOptions(lat=lat, lng=lng, map_type=map_type, zoom=zoom)
    p = gmap(apikey, gmap_options, title='Test Map', width=bokeh_width, height=bokeh_height)
    # definition of the column data source: 
    # source = ColumnDataSource(df)
    # see how we specify the x and y columns as strings, 
    # and how to declare as a source the ColumnDataSource:
    center = p.circle([lng], [lat], size=10, alpha=0.5, color='red')
    show(p)
    return p

In [54]:
p = plot(dt['lat'][0], dt['lng'][0])

That turned out great! Let's try it for some more points.

In [58]:
def plot_more(lat, lng, zoom=10, map_type='roadmap'):
    gmap_options = GMapOptions(lat=lat, lng=lng, map_type=map_type, zoom=zoom)
    p = gmap(apikey, gmap_options, title='Test Map', width=bokeh_width, height=bokeh_height)
    # definition of the column data source: 
    source = ColumnDataSource(dt)
    # see how we specify the x and y columns as strings, 
    # and how to declare as a source the ColumnDataSource:
    center = p.circle('lng', 'lat', size=10, alpha=0.5, color='red')
    show(p)
    return p

In [59]:
lat_c, lng_c = 37.625500, -111.439472

In [60]:
p = plot_more(lat_c, lng_c)

ERROR:bokeh.core.validation.check:E-1001 (BAD_COLUMN_NAME): Glyph refers to nonexistent column name. This could either be due to a misspelling or typo, or due to an expected column being missing. : key "x" value "lng", key "y" value "lat" [renderer: GlyphRenderer(id='2218', ...)]


In [61]:
dt

USGS        lat        long      alt                 basin  \
0    10140700  41.231819  111.984497  4285.00           Lower Weber   
1    10149000  40.118012  111.314622  6320.00          Spanish Fork   
2    10155200  40.554398  111.433243  5691.59                 Provo   
3    10224000  39.481898  112.393834  4660.00          Lower Sevier   
4    10308200  38.714627  119.764899  5400.00          Upper Carson   
..        ...        ...         ...      ...                   ...   
147  06892350  38.983337   94.964689   753.87  Lower Kansas, Kansas   
148  07169800  37.375597   96.185548   897.30                   Elk   
149  06775900  41.778611  100.525278  2798.18                Dismal   
150  06461500  42.902222  100.362222  2287.57       Middle Niobrara   
151  09076300  39.223200  106.857308  7575.00          Roaring Fork   

            lng  
0   -111.984497  
1   -111.314622  
2   -111.433243  
3   -112.393834  
4   -119.764899  
..          ...  
147  -94.964689  
148  -96.185548  
149 -100.525278  
150 -100.362222  
151 -106.857308  

[125 rows x 6 columns]

In [18]:
lat_t = dt['lat'].tolist()
long_t = dt['long'].tolist()
# lat_f = df['lat'].tolist()
# long_f = df['long'].tolist()

[41.23181944,
 40.11801206,
 40.55439805,
 39.48189754,
 38.7146274,
 39.53018865,
 39.7773722,
 48.5617222,
 48.9972222,
 48.20852,
 47.2744444,
 44.09888889,
 43.85855089,
 43.65405556,
 43.62088889,
 43.1961111,
 43.14305556,
 43.3508333,
 44.41666667,
 44.0697222,
 43.78194444,
 43.9272222,
 43.26277778,
 42.6247222,
 42.7675,
 42.52805556,
 42.6061111,
 42.6658333,
 41.9332368,
 41.8904614,
 42.7711111,
 43.25481389,
 42.26136944,
 43.66805556,
 43.3436111,
 43.66055556,
 44.08527778,
 44.29194444,
 44.11416667,
 43.9433333,
 45.2544444,
 44.2683333,
 45.1836111,
 44.7215194,
 44.9869444,
 45.75027778,
 46.00305556,
 46.08666667,
 46.08638889,
 46.84055556,
 46.4986111,
 46.4483333,
 39.55387495,
 38.9949943,
 38.6522188,
 38.4872189,
 38.25361396,
 39.37331729,
 39.14109959,
 39.1938333,
 40.00689167,
 37.8027744,
 40.93663889,
 39.7658209,
 39.75304299,
 37.8111077,
 37.7883333,
 37.59741667,
 36.74001667,
 36.3200333,
 35.8745,
 36.3188222,
 38.04332665,
 38.7841599,
 39.758876

In [30]:
# import gmplot

# gmap1 = gmplot.GoogleMapPlotter.from_geocode("Escalante, UT", 7, apikey='AIzaSyChKhMLyzg_dJYipHfbzo1QK7qbPW869WY')
# # coordinates of gages we are looking to predict
# gmap1.scatter(lat_t[0:1], long_t[0:1], color='red', size=10000, marker=False)
# gmap1.plot(lat_t[0:1], long_t[0:1], color='red', edge_width=5)

# # coordinates of gages we are using as features for predictions
# lat_f = df['lat'].tolist()
# long_f = df['long'].tolist()
# # gmap1.scatter(lat_f, long_f, color='blue', size=40, marker=True)

# # outputs into HTML file
# gmap1.draw('map.html')

In [27]:
help(gmap1.scatter)

Help on method scatter in module gmplot.gmplot:

scatter(lats, lngs, color=None, size=None, marker=True, c=None, s=None, symbol='o', **kwargs) method of gmplot.gmplot.GoogleMapPlotter instance
    Plot a collection of points.
    
    Args:
        lats ([float]): Latitudes.
        lngs ([float]): Longitudes.
    
    Optional:
    
    Args:
        color/c/edge_color/ec (str or [str]):
            Color of each point. Can be hex ('#00FFFF'), named ('cyan'), or matplotlib-like ('c'). Defaults to black.
        size/s (int or [int]): Size of each point, in meters (symbols only). Defaults to 40.
        marker (bool or [bool]): True to plot points as markers, False to plot them as symbols. Defaults to True.
        symbol (str or [str]): Shape of each point, as 'o', 'x', or '+' (symbols only). Defaults to 'o'.
        title (str or [str]): Hover-over title of each point (markers only).
        label (str or [str]): Label displayed on each point (markers only).
        precision (int or

In the next notebook:
    - pull USGS data for length that each gage is in operation
    - potentially cut time scale of data based on when the flow has never been above a certain threshold
    - each of the 125 gages should have a dataframe (list of dataframes)

In the notebook after that:
    - build 125 models (using Lasso Regression, we believe)
    - set threshold for accuracy in predicting flow
    - throw out models that do not meet that accuracy
    - throw out features in models that are no longer needed
    - do other prep for putting models into production (test in real time and have accuracy feedback)